src/script/descriptor.cpp

```diff
@@ -202,212 +204,236 @@ class BIP32PubkeyProvider final : public PubkeyProvider
-        if (!m_provider->ToPrivateString(arg, ret)) return false;
-        out = m_fn_name + "(" + std::move(ret) + ")";
-        return true;
-    }
-    bool Expand(int pos, const SigningProvider& arg, std::vector<CScript>& output_scripts, FlatSigningProvider& out) const override
-    {
-        CPubKey key;
-        KeyOriginInfo info;
-        if (!m_provider->GetPubKey(pos, arg, key, info)) return false;
-        output_scripts = std::vector<CScript>{m_script_fn(key)};
-        out.origins.emplace(key.GetID(), std::move(info));
-        out.pubkeys.emplace(key.GetID(), key);
-        return true;
-    }
-};
-
-CScript P2PKHGetScript(const CPubKey& pubkey) { return GetScriptForDestination(pubkey.GetID()); }
-CScript P2PKGetScript(const CPubKey& pubkey) { return GetScriptForRawPubKey(pubkey); }
-CScript P2WPKHGetScript(const CPubKey& pubkey) { return GetScriptForDestination(WitnessV0KeyHash(pubkey.GetID())); }
-
-/** A parsed multi(...) descriptor. */
-class MultisigDescriptor : public Descriptor
-{
-    int m_threshold;
-    std::vector<std::unique_ptr<PubkeyProvider>> m_providers;
-
-public:
-    MultisigDescriptor(int threshold, std::vector<std::unique_ptr<PubkeyProvider>> providers) : m_threshold(threshold), m_providers(std::move(providers)) {}
-
-    bool IsRange() const override
-    {
-        for (const auto& p : m_providers) {
-            if (p->IsRange()) return true;
+        for (const auto& pubkey : m_pubkeys) {
+            if (pubkey->IsRange()) return true;
+        }
+        if (m_script) {
+            if (m_script->IsRange()) return true;
         }
         return false;
     }
 
-    std::string ToString() const override
+    bool ToStringHelper(const SigningProvider* arg, std::string& out, bool priv) const
     {
-        std::string ret = strprintf("multi(%i", m_threshold);
-        for (const auto& p : m_providers) {
-            ret += "," + p->ToString();
+        std::string extra = ToStringExtra();
+        size_t pos = extra.size() > 0;
+        std::string ret = m_name + "(" + extra;
+        for (const auto& pubkey : m_pubkeys) {
+            if (pos++) ret += ",";
+            std::string tmp;
+            if (priv) {
+                if (!pubkey->ToPrivateString(*arg, tmp)) return false;
+            } else {
+                tmp = pubkey->ToString();
+            }
+            ret += std::move(tmp);
         }
-        return std::move(ret) + ")";
-    }
-
-    bool ToPrivateString(const SigningProvider& arg, std::string& out) const override
-    {
-        std::string ret = strprintf("multi(%i", m_threshold);
-        for (const auto& p : m_providers) {
-            std::string sub;
-            if (!p->ToPrivateString(arg, sub)) return false;
-            ret += "," + std::move(sub);
+        if (m_script) {
+            if (pos++) ret += ",";
+            std::string tmp;
+            if (!m_script->ToStringHelper(arg, tmp, priv)) return false;
+            ret += std::move(tmp);
```

Comment:
Same here? The result is passed as a const reference argument a the move is hence a noop?

Reply:
Fixed.